### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by sowecthal**
### Test Task

1. **Fork the following repository (make it public):**
   - [https://github.com/ganatan/angular-starter](https://github.com/ganatan/angular-starter)

2. **Create a `Dockerfile`** based on an Alpine image, utilizing Nginx as the web server.

3. **Create a GitHub Actions pipeline** that performs the following steps:
   - Run tests in the repository (if tests fail, send a failure notification and stop the execution of the pipeline).
   - Run `ng build`.
   - Create a Docker image based on your `Dockerfile` with `nginx.conf` and assets generated by the build step (`dist/angular-starter`) inside the container.
   - Store the created Docker image in GitHub Packages.
   - Send an email on success or failure to your personal email (e.g., if tests fail, send a "fail" notification).
   - _Bonus step_: Measure the time of the entire pipeline execution and include that information in the email notification.

4. **Create a `docker-compose.yaml` file** that will run the produced Docker image from GitHub Packages on your local PC. The output of `docker-compose up` followed by `curl localhost:8080` should display the full HTML of the built Angular application.

### Expected files

The following files should be added to the forked repository:
- `.github/workflows/ci.yaml` (GitHub Actions pipeline)
- `nginx.conf`
- `Dockerfile`
- `docker-compose.yaml`

### Security Requirements

- All processes inside the container must run under the "Nginx" user, with no "root" processes active inside the working container.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        with:
+          node-version: 18
+
+      - name: Install Clean dependencies
+        working-directory: ./angular
+        run: npm ci
+
+      - name: Run headless tests
+        id: run-tests
+        run: npm test -- --watch=false --browsers=ChromeHeadless
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Let's use corresponding `test` command from package.json?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+      - name: Install Clean dependencies
+        working-directory: ./angular
+        run: npm ci
+
+      - name: Run headless tests
+        id: run-tests
+        run: npm test -- --watch=false --browsers=ChromeHeadless 
+        working-directory: ./angular
+        continue-on-error: true
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Pls reorganize workflow using `always` status check function https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#status-check-functions

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        id: build-project
+        if: steps.run-tests.outcome == 'success'
+        run: npm run build
+        working-directory: ./angular
+        continue-on-error: true
+
+      - name: Log in to GitHub Container Registry  
+        id: login-registry
+        if: steps.build-project.outcome == 'success'
+        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login ghcr.io -u ${{ github.actor }} --password-stdin
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Can you define all secrets this workflow uses in `on` section? https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#on

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        continue-on-error: true
+
+      - name: Build Docker image
+        id: build-docker
+        if: steps.login-registry.outcome == 'success'
+        run: docker build -t ghcr.io/${{ github.actor }}/angular-react-starter-image .
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Let's add a separate input or env variable for docker registry, i.e. `ghcr.io/${{ github.actor }}/angular-react-starter-image` (or you can split it to 2 separate vars). So in case if we would like to change docker registry, we could do it in one place

Let's also add possibility to specify the tag we want, so docker image will have `latest` and custom tag if that is passed

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      - name: Get pipeline execution time
+        id: get-time
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: echo "execution_time=$( gh run list --workflow ci.yaml --limit 1 | grep -E --only-matching '([0-9]{1,2}[hms]){1,3}' )" >> $GITHUB_ENV
+
+      - name: Send email notification
+        uses: dawidd6/action-send-mail@v3
+        with:
+          server_address: smtp.gmail.com
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Let's pass smtp params also as vars (inputs or env vars in the beginning of the scipt)

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      - name: Checkout code from the repository
+        uses: actions/checkout@v4
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: 18
+
+      - name: Install Clean dependencies
+        working-directory: ./angular
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Can you set default working directory instead of repeating it for lots of steps?

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+RUN mkdir -p /var/cache/nginx/client_temp /var/run/nginx
+
+RUN chown -R nginx:nginx /var/cache/nginx /var/run/nginx /usr/share/nginx/html
+
+COPY ./nginx.conf /etc/nginx/nginx.conf
+
+COPY ./angular/dist/angular-starter/ /usr/share/nginx/html
+
+USER nginx
```
Comment (kseniyashaydurova) [inline Dockerfile]: Does your container start nginx by default?

File: docker-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+version: '3.8'
+
+services:
+  web:
+    image: ghcr.io/sowecthal/angular-react-starter-image:latest
+    ports:
+      - 8080:80
+    restart: always
```
Comment (kseniyashaydurova) [inline docker-compose.yaml:8]: Why restart is set to `always`?

File: nginx.conf [hunk 9]
```diff
@@ -0,0 +1,16 @@
+worker_processes auto;
+pid /var/run/nginx/nginx.pid;
+
+events {}
```
Comment (kseniyashaydurova) [inline nginx.conf:5]: Is it needed?

**Review by kseniyashaydurova — CHANGES_REQUESTED** [review]
Polina, good job, it really works! Can you pls show us proofs that emails are sent in successful and failed cases

**Comment by sowecthal** [comment]
**Failed case:**
failure

**Successful case:**
success

File: Dockerfile (hunk 7 above)
Comment (sowecthal) [reply Dockerfile]: Yes. Even though I didn't explicitly used an instruction to start Nginx (CMD/ENTRYPOINT), Nginx still starts by default due to the settings in the base Nginx Docker image

File: docker-compose.yaml (hunk 8 above)
Comment (sowecthal) [reply docker-compose.yaml:8]: This option is specifically needed for restarting in case the container suddenly stops (for example, if the program completes its execution normally or crashes)

File: nginx.conf (hunk 9 above)
Comment (sowecthal) [reply nginx.conf:5]: 1. If worker_processes is not specified, the default value of worker processes will be set to 1:
изображение 
With parameter auto it will be set to the number of available CPU cores:
изображение

2. User nginx doesn't have permission to write in the `/var/run/nginx.pid`. In my Dockerfile I set the user `nginx ` as the owner of the `/var/run/nginx` directory, and reassign the `pid `there.

3. Without `events {}` the Nginx webserver won't even start.

File: .github/workflows/ci.yaml (hunk 6 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 4 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: Dockerfile (hunk 7 above)
Comment (dmitry-mightydevops) [reply Dockerfile]: Polina your dockerfile should be explicit. What if the base image changes the command tomorrow and it fails? I.e change the location of the configuration file etc.

Explicit is always better than implicit. Good python rule, that we also follow in the team when we work on the solutions.

Comment (dmitry-mightydevops) [reply Dockerfile]: That helps us avoid the "assumptions".

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -0,0 +1,108 @@
+name: CI Pipeline
+
+on:
+  workflow_call:
+    secrets:
+      MAIL_USERNAME:
+        description: 'Username for the email notification step'
+        required: true
+      MAIL_PASSWORD:
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: prefer always lower_case naming. The environment vars named uppercase anywhere is like a sore in the eye.

Environment variables (like PATH, EDITOR, PAGER) are uppercase, and special shell variables (like UID, RANDOM, SECONDS) are uppercase. If you use uppercase for other purposes, you risk overriding such environment or special variables by accident. Especially if you're writing an sh script.

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -0,0 +1,108 @@
+name: CI Pipeline
+
+on:
+  workflow_call:
+    secrets:
+      MAIL_USERNAME:
+        description: 'Username for the email notification step'
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: all the values by default are strings, so you don't really need to use quotes.

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        required: true
+      MAIL_PASSWORD:
+        description: 'Password for the email notification step'
+        required: true
+    inputs:
+      custom_docker_tag:
+        description: 'Custom Docker tag'
+        # If not specified, then the 'latest' tag will be used
+        default: ''
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: then your default value should be set to latest
```suggestion
        # If not specified, then the 'latest' tag will be used
        default: latest
```

File: .github/workflows/ci.yaml [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+  DOCKER_REGISTRY: ghcr.io
+  DOCKER_REPO: angular-react-starter-image
+  DOCKER_TAG: ${{ inputs.custom_docker_tag || 'latest' }}
+  MAIL_SERVER_ADDRESS: smtp.gmail.com
+  MAIL_SERVER_PORT: 465
+
+jobs:
+
+  build:
+    runs-on: ubuntu-latest
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: make code consistent, so if you don't have empty rows in parent keys above, don't use it here. Or use it everywhere.

File: .github/workflows/ci.yaml [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        if: ${{ always() }}
+        run: npm run test -- --watch=false --browsers=ChromeHeadless 
+
+      - name: Build Angular project
+        id: build-project
+        if: steps.run-tests.outcome == 'success'
+        run: npm run build
+
+      - name: Log in to GitHub Container Registry  
+        id: login-registry
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: have a consistent naming
- Camelcase some other words
- Camelcase Some Other Words

but don't mix. This is like sore in the eye as well.

File: .github/workflows/ci.yaml [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        if: steps.build-docker.outcome == 'success'
+        run: docker push $DOCKER_REGISTRY/${{ github.actor }}/$DOCKER_REPO:$DOCKER_TAG
+
+      - name: Get pipeline execution time
+        id: get-time
+        if: ${{ always() }}
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: echo "execution_time=$( gh run list --workflow ci.yaml --limit 1 | grep -E --only-matching '([0-9]{1,2}[hms]){1,3}' )" >> $GITHUB_ENV
+
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: - use multiline script instead a single oneliner
- gh has json output - use it and query for the desired fields, then calculation the duration

if gh changes the output and change notations for duration or remove that field completely - your solution will stop working.

File: .github/workflows/ci.yaml [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+          server_address: ${{ env.MAIL_SERVER_ADDRESS }}
+          server_port: $MAIL_SERVER_PORT
+
+          username: ${{ secrets.MAIL_USERNAME }}
+          password: ${{ secrets.MAIL_PASSWORD }}
+
+          subject: GITHUB ACTIONS JOB RESULT
+
+          to: "${{ github.actor_id }}+${{ github.actor }}@users.noreply.github.com"
+          from: GitHub Actions Pipeline
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: don't use empty rows

File: Dockerfile (hunk 7 above)
Comment (dmitry-mightydevops) [inline Dockerfile]: Makse condensed. Separate with lines ONLY when you need to emphasize something. Here you emphasize everything, means no emphasis (as everything is equal)

File: nginx.conf [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+
+http {
+    server {
+        listen 80;
```
Comment (dmitry-mightydevops) [inline nginx.conf]: Ports Below 1024: On Unix-like systems, ports below 1024 are considered "privileged" ports.
So it's better to use something like 8080

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml [hunk 18]
```diff
@@ -1,17 +1,43 @@
 name: CI Pipeline
 
 on:
+  workflow_call:
+    secrets:
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Why there is no `GITHUB_TOKEN`?

File: .github/workflows/ci.yaml [hunk 19]
```diff
@@ -1,17 +1,43 @@
...
     branches:
       - master
   pull_request:
     branches:
       - master
 
+env:
+  DOCKER_REGISTRY: ghcr.io
+  DOCKER_REPO: angular-react-starter-image
+  DOCKER_TAG: ${{ inputs.custom_docker_tag || 'latest' }}
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: LEt's not add such a var, let's pass a default value to input instead

File: .github/workflows/ci.yaml [hunk 20]
```diff
@@ -1,17 +1,43 @@
...
+  MAIL_SERVER_ADDRESS: smtp.gmail.com
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Let's move `MAIL_SERVER_ADDRESS` and `MAIL_SERVER_PORT` to inputs, cause basically these values could be changed (as well as docker registry, repo). 

All these values could be different, for example today you use Gmail, tomorrow you would like to configure SMTP on AWS SES or Sendgrid. It is much better to allow to pass these variables as inputs for your workflow, instead of changing workfow itself. It is also true for docker registry and repo. Today you use gh, tomorrow you will use AWS ECR or docker registry and would like to change repo name

File: .github/workflows/ci.yaml [hunk 21]
```diff
@@ -22,51 +48,47 @@ jobs:
...
       - name: Install Clean dependencies
-        working-directory: ./angular
         run: npm ci
 
       - name: Run headless tests
         id: run-tests
-        run: npm test -- --watch=false --browsers=ChromeHeadless 
-        working-directory: ./angular
-        continue-on-error: true
+        if: ${{ always() }}
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: You can use just `always()` everywhere. 

Btw, I don't think that you need to add `always` to tests, because basically it makes no sense to run tests if you got errors on `install clean dependencies` step for example. So Let's leave it without any condition.

File: .github/workflows/ci.yaml [hunk 22]
```diff
@@ -22,51 +48,47 @@ jobs:
...
+        run: npm run test -- --watch=false --browsers=ChromeHeadless
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml:80]: Why did you add these args to tests?

File: .github/workflows/ci.yaml [hunk 23]
```diff
@@ -22,51 +48,47 @@ jobs:
...
         id: run-tests
-        run: npm test -- --watch=false --browsers=ChromeHeadless 
-        working-directory: ./angular
-        continue-on-error: true
+        if: ${{ always() }}
+        run: npm run test -- --watch=false --browsers=ChromeHeadless 
 
       - name: Build Angular project
         id: build-project
         if: steps.run-tests.outcome == 'success'
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Try to not use these `steps.run-tests.outcome` conditions checks everywhere and check how it will behave. 

Basically the suggestion with `always` was about `Get pipeline execution time` and `Send email notification` steps. I.e. everything should run without condition by default in the same way as task suggest. I.e. workflow will fail on a first error (that is ok and ok for tests too), but `Get pipeline execution time` and `Send email notification` steps would run always.

File: .github/workflows/ci.yaml [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+
+      - name: Build Angular project
+        id: build-project
+        if: steps.run-tests.outcome == 'success'
+        run: npm run build
+
+      - name: Log in to GitHub Container Registry  
+        id: login-registry
+        if: steps.build-project.outcome == 'success'
+        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login $DOCKER_REGISTRY -u ${{ github.actor }} --password-stdin
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Use multiline instead of single-liner

File: .github/workflows/ci.yaml [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+      - name: Build Docker image
+        id: build-docker
+        if: steps.login-registry.outcome == 'success'
+        run: docker build -t $DOCKER_REGISTRY/${{ github.actor }}/$DOCKER_REPO:$DOCKER_TAG .
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Define this var ass env var and reuse it everywhere where it repeats? I.e. `DOCKER_REGISTRY_URL=<some calc based on inputs>`

File: .github/workflows/ci.yaml [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: echo "execution_time=$( gh run list --workflow ci.yaml --limit 1 | grep -E --only-matching '([0-9]{1,2}[hms]){1,3}' )" >> $GITHUB_ENV
+
+      - name: Send email notification
+        if: ${{ always() }}
+        uses: dawidd6/action-send-mail@v3
+        with:
+          server_address: ${{ env.MAIL_SERVER_ADDRESS }}
+          server_port: $MAIL_SERVER_PORT
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Why do you pass `MAIL_SERVER_ADDRESS` in absolutely different way with `MAIL_SERVER_PORT`? Pls always use the same syntax for similar things

File: .github/workflows/ci.yaml [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+
+          username: ${{ secrets.MAIL_USERNAME }}
+          password: ${{ secrets.MAIL_PASSWORD }}
+
+          subject: GITHUB ACTIONS JOB RESULT
+
+          to: "${{ github.actor_id }}+${{ github.actor }}@users.noreply.github.com"
```
Comment (kseniyashaydurova) [inline .github/workflows/ci.yaml]: Let's move this `to` email to a var too, if nothing is passed it would be calculated in a way like it is now

**Review by kseniyashaydurova — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 22 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml:80]: `--watch=false` disables the watch mode, which is unnecessary in CI/CD environments where tests should run only once.
`--browsers=ChromeHeadless` runs tests in a headless Chrome browser, which is ideal, because it doesn't require a graphical interface.

File: .github/workflows/ci.yaml (hunk 18 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Fixed

File: .github/workflows/ci.yaml (hunk 19 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 20 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 21 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 23 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Thank you. Done

File: .github/workflows/ci.yaml (hunk 24 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 25 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done. Named it IMAGE_NAME

File: .github/workflows/ci.yaml (hunk 26 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Fixed

File: .github/workflows/ci.yaml (hunk 27 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 15 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml (hunk 12 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Fixed

File: .github/workflows/ci.yaml (hunk 11 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Fixed

File: Dockerfile (hunk 7 above)
Comment (sowecthal) [reply Dockerfile]: Thank you. Fixed

File: .github/workflows/ci.yaml [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+      run:
+        working-directory: ./angular
+    steps:
+      - name: Checkout code from the repository
+        uses: actions/checkout@v4
+
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: 18
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: versions should never be hardcoded, pass them from the inputs please - that helps maintenance as every config is in one place.

File: .github/workflows/ci.yaml [hunk 29]
```diff
@@ -0,0 +1,121 @@
...
+
+      - name: Install clean dependencies
+        run: npm ci
+
+      - name: Run headless tests
+        id: run-tests
+        run: npm run test -- --watch=false --browsers=ChromeHeadless 
+
+      - name: Build Angular project
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: pls see my previous comments about naming the steps - let's be clean and consistent everywhere

File: .github/workflows/ci.yaml [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+        id: build-project
+        run: npm run build
+
+      - name: Log in to GitHub container registry  
+        id: login-registry
+        run: | 
+          echo "${{ secrets.github_token }}" | docker login \
+          ${{ inputs.docker_registry }} -u ${{ github.actor }} \
+          --password-stdin
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]:
```suggestion
          echo "${{ secrets.github_token }}" \
          | docker login ${{ inputs.docker_registry }} -u ${{ github.actor }} --password-stdin
```

File: .github/workflows/ci.yaml [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+        run: docker push $IMAGE_NAME
+
+      - name: Get pipeline execution time
+        id: get-time
+        if: always()
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          start_time=$(gh run list --json createdAt -w ci.yaml -L 1 | jq -r '.[0].createdAt')
+          end_time=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: use updatedAt, of the gh output to get end_time

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 31 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: I'm not sure if I understood you correctly, but I changed `end_time` to use `updatedAt` instead of the current time. The problem is that `updatedAt` is supposed to be the time of the last workflow status update, but it's lagging behind the current time

File: .github/workflows/ci.yaml (hunk 28 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Thank you. Done

File: .github/workflows/ci.yaml [hunk 32]
```diff
@@ -0,0 +1,125 @@
...
+        run: docker push $IMAGE_NAME
+
+      - name: Get pipeline execution time
+        id: get-time
+        if: always()
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          start_time=$(gh run list --json createdAt -w ci.yaml -L 1 | jq -r '.[0].createdAt')
+          end_time=$(gh run list --json updatedAt -w ci.yaml -L 1 | jq -r '.[0].updatedAt')
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: here you make double API calls. Let's reduce to one.

File: .github/workflows/ci.yaml [hunk 33]
```diff
@@ -0,0 +1,125 @@
...
+
+      - name: Send email notification
+        if: always()
+        uses: dawidd6/action-send-mail@v3
+        with:
+          server_address: ${{ inputs.mail_server_address }}
+          server_port: ${{ inputs.mail_server_port }}
+          username: ${{ secrets.mail_username }}
+          password: ${{ secrets.mail_password }}
+          subject: GITHUB ACTIONS JOB RESULT
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: uppercase is like a sore in the eye. Better to extract this into configuration var (*i.e. inputs) and consume like you do with other attrs of the task.

File: Dockerfile [hunk 34]
```diff
@@ -0,0 +1,11 @@
+FROM nginx:stable-alpine
```
Comment (dmitry-mightydevops) [inline Dockerfile]: - PIN via sha256 hash to specific version
  - Add comment above showing that the tag was mapped to stable-alpine 
- Add link to the tag in the docker registry, from which you pull the image

The reason you want to do that is that builds should produce the same image all the time. i.e. reproducable builds.

In your case image you will build now in 2 months later will be different, which is bad for maintenance and trobleshooting.

File: Dockerfile [hunk 35]
```diff
@@ -0,0 +1,11 @@
...
+
+RUN mkdir -p /var/cache/nginx/client_temp /var/run/nginx
+RUN chown -R nginx:nginx /var/cache/nginx /var/run/nginx /usr/share/nginx/html
+
+COPY ./nginx.conf /etc/nginx/nginx.conf
+COPY ./angular/dist/angular-starter/ /usr/share/nginx/html
+
+USER nginx
+
+CMD ["nginx", "-g", "daemon off;"]
```
Comment (dmitry-mightydevops) [inline Dockerfile]: configuration of running nginx should rely on config file i.e. nginx.conf, not arguments that you pass in CMD. Otherwise you reduce your maintainability. Your nginx config file should be the ONLY source of truth.

The more options you spread in different files - the more complex the maintenance will be in the future and will require more time.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 31 above)
Comment (dmitry-mightydevops) [reply .github/workflows/ci.yaml]: so they update the status and then set the updateTime, so we can use it to calculate the duration. It's enough for checking the speed of the pipeline.

File: .github/workflows/ci.yaml (hunk 32 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Fixed

File: .github/workflows/ci.yaml (hunk 33 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: Dockerfile (hunk 34 above)
Comment (sowecthal) [reply Dockerfile]: Done

File: Dockerfile (hunk 35 above)
Comment (sowecthal) [reply Dockerfile]: Done

File: .github/workflows/ci.yaml [hunk 36]
```diff
@@ -0,0 +1,131 @@
...
+        id: get-time
+        if: always()
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          run_info=$(gh run list --json createdAt,updatedAt -w ci.yaml -L 1)
+          start_time=$(echo "$run_info" | jq -r '.[0].createdAt')
+          end_time=$(echo "$run_info" | jq -r '.[0].updatedAt')
+          duration=$(($(date -d "$end_time" +%s) - $(date -d "$start_time" +%s)))
+          echo "execution_time=${duration}s" >> $GITHUB_ENV
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml]: let's simplify this and make it oneliner.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
also squash/rebase all your existing commits into one, so you have small history

File: .github/workflows/ci.yaml (hunk 36 above)
Comment (sowecthal) [reply .github/workflows/ci.yaml]: Done

File: .github/workflows/ci.yaml [hunk 37]
```diff
@@ -0,0 +1,126 @@
...
+      - name: Push docker image
+        id: push-docker
+        run: docker push $IMAGE_NAME
+
+      - name: Get pipeline execution time
+        id: get-time
+        if: always()
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: echo "execution_time=$(gh run list --json createdAt,updatedAt -w ci.yaml -L 1 | jq -r '.[0] | (( (strtotime(.updatedAt) - strtotime(.createdAt)) ))')s" >> $GITHUB_ENV
```
Comment (dmitry-mightydevops) [inline .github/workflows/ci.yaml:106]: could be simplified to have less parenthesis:
```
DURATION=$(gh run list --json createdAt,updatedAt | jq -r '.[0] | (.updatedAt | fromdateiso8601) - (.createdAt | fromdateiso8601)')
```

**Review by dmitry-mightydevops — APPROVED** [review]
**OK all good! Nice work :)** 



# See you in the camp!

**Comment by dmitry-mightydevops** [comment]
@kseniyashaydurova is on the vacation next 2 weeks so I approve for her as well :)

**Review by kseniyashaydurova — APPROVED** [review]
